client/client.go

```diff
@@ -48,6 +48,23 @@ var DefaultTransport CancelableTransport = &http.Transport{
 	TLSHandshakeTimeout: 10 * time.Second,
 }
 
+type EndpointSelectionMode int
+
+const (
+	// EndpointSelectionDefault is to pick an endpoint in a random manner.
+	EndpointSelectionDefault EndpointSelectionMode = iota
+
+	// EndpointSelectionPrioritizeLeader is to prioritize leader for reducing needless
+	// forward between follower and leader.
+	//
+	// This mode should be used with Client.AutoSync().
+	EndpointSelectionPrioritizeLeader
+
+	// EndpointSelectionPreference is to manually picks an endpoint
+	// with using Config.PreferenceIndex as an index.
+	EndpointSelectionPreference
```

Comment:
@xiang90 do you mean boolean flag like the previous PR ( https://github.com/coreos/etcd/pull/3950 ) is enough for now?